Add custom CosmJs elements to my own chain (#998)
cosmos/sdk-tutorials: academy/4-my-own-chain/cosmjs-gui.md

        const creator = (await offlineSigner.getAccounts())[0].address
        const client: CheckersSigningStargateClient = await CheckersSigningStargateClient.connectWithSigner(
            this.props.rpcUrl,
            offlineSigner,
            {
                gasPrice: GasPrice.fromString("1stake"),
            },
        )
        this.setState({ creator: creator, signingClient: client })
        return { creator: creator, signingClient: client }
    }
    ```

    Setting up Keplr is idempotent, so repeating these operations more than once is harmless. You may want to separate these actions into more defined methods at a later optimization stage.

    Note too that a default gas price is passed in, so that you can use `"auto"` when sending a transaction.

Your component is now ready to send transactions to the blockchain. Why choose `NewGameModal.tsx` as the example? Because this is where a new game is created. 

## Create a new game

This modal window pops up when you click on <kbd>New Game</kbd>:

![Modal window to create game](/window-create-game.png)

The player names look ready-made to take the player's Cosmos addresses.

> Look around the code and you see that the action takes place in `src/components/Menu/NewGameModal/NewGameModal.tsx`. Thanks to the previous preparation with `CheckersSigningStargateClient` it is ready to send transactions>

Author: ```suggestion
Look around the code and you see that the action takes place in `src/components/Menu/NewGameModal/NewGameModal.tsx`. Thanks to the previous preparation with `CheckersSigningStargateClient` it is ready to send transactions:
```

<HighlightBox type="info">

> Could an AI player and the blockchain mix together? If _the AI_ has access to a private key that lets it send transactions, it would be a bona fide player. This could be implemented with backend scripts running on a server. See [Checkers backend scripts](./server-side.md) for an example of backend scripts with a different purpose.

Author: ```suggestion
Could an AI player and the blockchain mix together? If _the AI_ has access to a private key that lets it send transactions, it would be a bona fide player. This could be implemented with backend scripts running on a server. See [Checkers backend scripts](./server-side.md) for an example of backend scripts for a different use-case.
```
@AndrewB9lab

Reviewer: Is good.

   }
    ```

    Note that this creates it only once by saving the client in the state.

Instantiating a `CheckersSigningStargateClient` is more involved, as you will see later on.

First, make use of what you already prepared. Why choose to start with `MenuContainer`? Because that is where all the games are listed.

## Show all games

To show all games, you only need the read-only `StargateClient`. There is no need to worry about private keys for now. Where do you query for the games?

### All games container

Look into `src/components/Menu/MenuContainer.tsx` for the `componentDidMount` method, which accesses the browser storage for saved games. You can replace the storage logic with the following steps:

* Obtain a `StargateClient` with the `getStargateClient` method.
* Obtain the blockchain games (without pagination).
* Convert the blockchain games to the format the component expects.
* Save them in `saved: IGameInfo[]` of the component's state.

A good next step is to add a new function to `CheckersStargateClient` to handle the call and the conversion, or to add a helper function that takes a client as a parameter.

### A specific GUI method

> In the future you may want to reuse the `CheckersStargateClient` code in another GUI or for backend scripts. To avoid polluting the code, and to avoid a helper where you need to pass a Stargate client as parameter, add [an extension method](https://www.c-sharpcorner.com/article/learn-about-extension-methods-in-typescript/#:~:text=Extension%2Dmethod%20gives%20you%20the,any%20data%2Dtype%20you%20want) to `CheckersStargateClient`.

Author: ```suggestion
In the future you may want to reuse the `CheckersStargateClient` code in another GUI or for backend scripts. To avoid polluting the code, and to avoid a less-elegant helper where you need to pass a Stargate client as parameter, add [an extension method](https://www.c-sharpcorner.com/article/learn-about-extension-methods-in-typescript/#:~:text=Extension%2Dmethod%20gives%20you%20the,any%20data%2Dtype%20you%20want) to `CheckersStargateClient`.
```

In a new `src/types/checkers/extensions-gui.ts`:

> 1. Declare the new extension method to your `CheckersStargateClient` _module_:

Author: ```suggestion
1. Declare the new extension method that extends your `CheckersStargateClient` _module_:
```

Reviewer: Is good.

    }
    ```

    Note:

    * You force `isSaved` to `true` since it is always saved in the blockchain.
    * Saving `game.index` to state is unimportant because it is actually passed in `IGameContainerProps.index`.
    * By having a separate `loadGame`, you can call it again if you know the game has changed.

Restart `npm start` and you should now see your game. If you have access to `checkersd`, or any other way to send a transaction, you can make a move from the command line and refresh to confirm the change. If not, wait for the procedure about playing.

![Game with one move played](/show-game-1-move.png)

## Integrate with Keplr

> So far you have _only_ made it possible to show the state of games and of the blockchain. This allows your users poke around without unnecessarily asking them to connect. However, to create a game or play in one, you need to make transactions. This is where you need to make integration with Keplr possible.

Author: ```suggestion
So far you have _only_ made it possible to show the state of games and of the blockchain. This allows your users to poke around without unnecessarily asking them to connect. However, to create a game or play in one, you need to make transactions. This is where you need to make integration with Keplr possible.
```
Not sure it is missing "to" or not. @AndrewB9lab

Reviewer: It certainly is, thank you for the catch!

The player names look ready-made to take the player's Cosmos addresses.

> Look around the code and you see that the action takes place in `src/components/Menu/NewGameModal/NewGameModal.tsx`.Thanks to the previous preparation with `CheckersSigningStargateClient` it is ready to send transactions>

Author: ```suggestion
Look around the code and you see that the action takes place in `src/components/Menu/NewGameModal/NewGameModal.tsx`.Thanks to the previous preparation with `CheckersSigningStargateClient` it is ready to send transactions:
```

    maxLength={45}
    ```

4. Back in `NewGameModal.tsx`, change the declaration of `handleSubmit` and make it `async`:

    ```typescript [https://github.com/cosmos/academy-checkers-ui/blob/f6a96b7/src/components/Menu/NewGameModal/NewGameModal.tsx#L127]
    private async handleSubmit(event: any): Promise<void>
    ```

> 5. In `handleSubmit`, avert the need to save to local storage:

Author: ```suggestion
5. In `handleSubmit`, avert the need to save to local storage and call your create game extension method:
```

    <HighlightBox type="tip">

    There is a potentially hard-to-reproduce-in-production **race condition** bug here. The `loadGame` is done immediately after the transaction has completed. However, depending on the implementation of the RPC end point, the `playGuiMove` and `loadGame` calls may hit two different servers on the backend. In some instances, the server that answers your `loadGame` may not have fully updated its store and may in fact serve you the **old** version of your game.

>     Aas your GUI matures, you may want to show the _expected_ state of the game before you eventually show its _finalized_ state. Sometimes you may want to show the expected state of the game even before the transaction has completed, and add visual cues hinting at the fact it is a **provisional** state.

Author: ```suggestion
    As your GUI matures, you may want to show the _expected_ state of the game before you eventually show its _finalized_ state. Sometimes you may want to show the expected state of the game even before the transaction has completed, and add visual cues hinting at the fact it is a **provisional** state.
```
@AndrewB9lab

Reviewer: Thank you...

    The same can happen when creating a game, where the second server may return `null` if it has not been updated yet.

    </HighlightBox>

> Now you can play test in the GUI. Make sure to put your Keplr address as the black player, so that you start with this one. You will need a second account on Keplr with which to play red, otherwise you must play red from the command line.

Author: ```suggestion
Now you can play test in the GUI. Make sure to put your Keplr address as the black player, so that you start with this one. You will need a second account on Keplr with which to play red, otherwise you must play red from the command line. Alternatively, put the same Keplr address for both black and red.
```
